tools/clang/lib/SPIRV/LowerTypeVisitor.cpp

```diff
@@ -8,27 +8,228 @@
 //===----------------------------------------------------------------------===//
 
 #include "LowerTypeVisitor.h"
-
 #include "clang/AST/Attr.h"
 #include "clang/AST/DeclCXX.h"
 #include "clang/AST/HlslTypes.h"
 #include "clang/SPIRV/AstTypeProbe.h"
+#include "clang/SPIRV/SpirvFunction.h"
+
+namespace {
+/// Returns the :packoffset() annotation on the given decl. Returns nullptr if
+/// the decl does not have one.
+hlsl::ConstantPacking *getPackOffset(const clang::NamedDecl *decl) {
+  for (auto *annotation : decl->getUnusualAnnotations())
+    if (auto *packing = llvm::dyn_cast<hlsl::ConstantPacking>(annotation))
+      return packing;
+  return nullptr;
+}
+
+/// Rounds the given value up to the given power of 2.
+inline uint32_t roundToPow2(uint32_t val, uint32_t pow2) {
+  assert(pow2 != 0);
+  return (val + pow2 - 1) & ~(pow2 - 1);
+}
+
+} // end anonymous namespace
 
 namespace clang {
 namespace spirv {
 
+bool LowerTypeVisitor::visit(SpirvFunction *fn, Phase phase) {
+  if (phase == Visitor::Phase::Init) {
+    // Lower the function return type.
+    const SpirvType *spirvReturnType =
+        lowerType(fn->getAstReturnType(), SpirvLayoutRule::Void,
+                  /*isRowMajor*/ llvm::None,
+                  /*SourceLocation*/ {});
+    fn->setReturnType(const_cast<SpirvType *>(spirvReturnType));
+
+    // Lower the SPIR-V function type if necessary.
+    fn->setFunctionType(const_cast<SpirvType *>(
+        lowerType(fn->getFunctionType(), SpirvLayoutRule::Void,
+                  fn->getSourceLocation())));
+  }
+  return true;
+}
+
+bool LowerTypeVisitor::visitInstruction(SpirvInstruction *instr) {
+  const QualType astType = instr->getAstResultType();
+  const SpirvType *hybridType = instr->getResultType();
+
+  // Lower QualType to SpirvType
+  if (astType != QualType({})) {
```

Comment:
Can just do `!astType`? Also other places.

Reply:
apparently can't do this.